File: archive/cdktf-ts/Pr6107/lib/compute-stack.ts
import { Construct } from 'constructs';
import { LaunchTemplate } from '@cdktf/provider-aws/lib/launch-template';
import { AutoscalingGroup } from '@cdktf/provider-aws/lib/autoscaling-group';
import { SecurityGroup } from '@cdktf/provider-aws/lib/security-group';
import { SecurityGroupRule } from '@cdktf/provider-aws/lib/security-group-rule';
import { DataAwsAmi } from '@cdktf/provider-aws/lib/data-aws-ami';
import { AwsProvider } from '@cdktf/provider-aws/lib/provider';

export interface ComputeStackProps {
  environmentSuffix: string;
  vpcId: string;
  privateSubnetIds: string[];
  targetGroupArn: string;
  instanceProfileName: string;
  region: string;
  provider?: AwsProvider;
}

export interface ComputeStackOutputs {
  asgName: string;
  launchTemplateId: string;
}

export class ComputeStack extends Construct {
  public readonly outputs: ComputeStackOutputs;

  constructor(scope: Construct, id: string, props: ComputeStackProps) {
    super(scope, id);

    const {
      environmentSuffix,
      vpcId,
      privateSubnetIds,
      targetGroupArn,
      instanceProfileName,
      region,
      provider,
    } = props;

    const drRole = region === 'us-east-1' ? 'primary' : 'dr';

    const commonTags = {
      Environment: environmentSuffix,
      CostCenter: 'payment-processing',
      'DR-Role': drRole,
      ManagedBy: 'cdktf',
    };

    // Get latest Amazon Linux 2 AMI
    const ami = new DataAwsAmi(this, `ami-${region}`, {
      mostRecent: true,
      owners: ['amazon'],
      filter: [
        {
          name: 'name',
          values: ['amzn2-ami-hvm-*-x86_64-gp2'],
        },
        {
          name: 'virtualization-type',
          values: ['hvm'],
        },
      ],
      provider: provider,
    });

    // Security Group for EC2 instances
    const instanceSg = new SecurityGroup(this, `instance-sg-${region}`, {
      name: `payment-instance-sg-${environmentSuffix}-${region}`,
      description: 'Security group for payment processing instances',
      vpcId: vpcId,
      tags: {
        ...commonTags,
        Name: `payment-instance-sg-${environmentSuffix}-${region}`,
      },
      provider: provider,
    });

    new SecurityGroupRule(this, `instance-http-ingress-${region}`, {
      type: 'ingress',
      fromPort: 80,
      toPort: 80,
      protocol: 'tcp',
      cidrBlocks: ['10.0.0.0/8'],
      securityGroupId: instanceSg.id,
      description: 'HTTP access from ALB',
      provider: provider,
    });

    new SecurityGroupRule(this, `instance-https-ingress-${region}`, {
      type: 'ingress',
      fromPort: 443,
      toPort: 443,
      protocol: 'tcp',
      cidrBlocks: ['10.0.0.0/8'],
      securityGroupId: instanceSg.id,
      description: 'HTTPS access from ALB',
      provider: provider,
    });

    new SecurityGroupRule(this, `instance-egress-${region}`, {
      type: 'egress',
      fromPort: 0,
      toPort: 0,
      protocol: '-1',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: instanceSg.id,
      description: 'Allow all outbound traffic',
      provider: provider,
    });

    // User data script
    const userData = `#!/bin/bash
yum update -y
yum install -y httpd
systemctl start httpd
systemctl enable httpd
echo "<h1>Payment Processing - ${region} - ${environmentSuffix}</h1>" > /var/www/html/index.html
echo "<p>Region: ${region}</p>" >> /var/www/html/index.html
echo "<p>DR Role: ${drRole}</p>" >> /var/www/html/index.html
`;

    // Launch Template
    const launchTemplate = new LaunchTemplate(
      this,
      `launch-template-${region}`,
      {
        name: `payment-lt-${environmentSuffix}-${region}`,
        imageId: ami.id,
        instanceType: 't3.large',
        iamInstanceProfile: {
          name: instanceProfileName,
        },
        vpcSecurityGroupIds: [instanceSg.id],
        userData: Buffer.from(userData).toString('base64'),
        blockDeviceMappings: [
          {
            deviceName: '/dev/xvda',
            ebs: {
              volumeSize: 20,
              volumeType: 'gp3',
              encrypted: 'true',
              deleteOnTermination: 'true',
            },
          },
        ],
        metadataOptions: {
          httpEndpoint: 'enabled',
          httpTokens: 'required',
          httpPutResponseHopLimit: 1,
        },
        monitoring: {
          enabled: true,
        },
        tagSpecifications: [
          {
            resourceType: 'instance',
            tags: {
              ...commonTags,
              Name: `payment-instance-${environmentSuffix}-${region}`,
            },
          },
          {
            resourceType: 'volume',
            tags: {
              ...commonTags,
              Name: `payment-volume-${environmentSuffix}-${region}`,
            },
          },
        ],
        provider: provider,
      }
    );

    // Auto Scaling Group
    const asg = new AutoscalingGroup(this, `asg-${region}`, {
      name: `payment-asg-${environmentSuffix}-${region}`,
      minSize: 2,
      maxSize: 6,
      desiredCapacity: 2,
      vpcZoneIdentifier: privateSubnetIds,
      targetGroupArns: [targetGroupArn],
      healthCheckType: 'ELB',
      healthCheckGracePeriod: 300,
      launchTemplate: {
        id: launchTemplate.id,
        version: '$Latest',
      },
      tag: [
        {
          key: 'Name',
          value: `payment-instance-${environmentSuffix}-${region}`,
          propagateAtLaunch: true,
        },
        {
          key: 'Environment',
          value: environmentSuffix,
          propagateAtLaunch: true,
        },
        {
          key: 'CostCenter',
          value: 'payment-processing',
          propagateAtLaunch: true,
        },
        {
          key: 'DR-Role',
          value: drRole,
          propagateAtLaunch: true,
        },
      ],
      provider: provider,
    });

    this.outputs = {
      asgName: asg.name,
      launchTemplateId: launchTemplate.id,
    };
  }
}

File: archive/cdktf-py/Pr6435/lib/compute_stack.py
"""Compute Stack - ALB, Auto Scaling Group, Launch Template."""

import base64
from constructs import Construct
from cdktf_cdktf_provider_aws.lb import Lb
from cdktf_cdktf_provider_aws.lb_target_group import LbTargetGroup, LbTargetGroupHealthCheck
from cdktf_cdktf_provider_aws.lb_listener import LbListener, LbListenerDefaultAction
from cdktf_cdktf_provider_aws.wafv2_web_acl_association import Wafv2WebAclAssociation
from cdktf_cdktf_provider_aws.launch_template import (
    LaunchTemplate,
    LaunchTemplateIamInstanceProfile,
    LaunchTemplateMetadataOptions,
    LaunchTemplateMonitoring,
)
from cdktf_cdktf_provider_aws.autoscaling_group import (
    AutoscalingGroup,
    AutoscalingGroupLaunchTemplate,
    AutoscalingGroupTag,
)
from cdktf_cdktf_provider_aws.data_aws_ami import DataAwsAmi, DataAwsAmiFilter
from cdktf_cdktf_provider_aws.data_aws_acm_certificate import DataAwsAcmCertificate


class ComputeStack(Construct):
    """Compute infrastructure for Node.js API servers."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        environment_suffix: str,
        vpc_id: str,
        public_subnet_ids: list,
        private_subnet_ids: list,
        alb_security_group_id: str,
        api_security_group_id: str,
        instance_profile_arn: str,
        waf_web_acl_id: str,
        aws_region: str,
        **kwargs
    ):
        """Initialize compute stack."""
        super().__init__(scope, construct_id)

        # Application Load Balancer
        self._alb = Lb(
            self,
            "alb",
            name=f"payment-alb-{environment_suffix}",
            internal=False,
            load_balancer_type="application",
            security_groups=[alb_security_group_id],
            subnets=public_subnet_ids,
            enable_deletion_protection=False,
            enable_http2=True,
            tags={
                "Name": f"payment-alb-{environment_suffix}",
            },
        )

        # Target Group
        self._target_group = LbTargetGroup(
            self,
            "target_group",
            name=f"payment-tg-{environment_suffix}",
            port=3000,
            protocol="HTTP",
            vpc_id=vpc_id,
            target_type="instance",
            deregistration_delay="30",
            health_check=LbTargetGroupHealthCheck(
                enabled=True,
                healthy_threshold=2,
                unhealthy_threshold=3,
                timeout=5,
                interval=30,
                path="/health",
                protocol="HTTP",
                matcher="200",
            ),
            tags={
                "Name": f"payment-tg-{environment_suffix}",
            },
        )

        # ALB Listener (HTTPS would require ACM certificate)
        # For demo purposes, using HTTP listener
        # In production, use HTTPS with ACM certificate
        LbListener(
            self,
            "alb_listener",
            load_balancer_arn=self._alb.arn,
            port=80,
            protocol="HTTP",
            default_action=[
                LbListenerDefaultAction(
                    type="forward",
                    target_group_arn=self._target_group.arn,
                )
            ],
        )

        # Associate WAF with ALB
        Wafv2WebAclAssociation(
            self,
            "waf_alb_association",
            resource_arn=self._alb.arn,
            web_acl_arn=waf_web_acl_id,
        )

        # Get latest Amazon Linux 2023 AMI
        ami = DataAwsAmi(
            self,
            "amazon_linux_2023",
            most_recent=True,
            owners=["amazon"],
            filter=[
                DataAwsAmiFilter(
                    name="name",
                    values=["al2023-ami-*-x86_64"],
                ),
                DataAwsAmiFilter(
                    name="virtualization-type",
                    values=["hvm"],
                ),
            ],
        )

        # User Data Script for Node.js API
        user_data = f"""#!/bin/bash
# Update system
dnf update -y

# Install Node.js 18
dnf install -y nodejs npm

# Install CloudWatch agent
dnf install -y amazon-cloudwatch-agent

# Create app directory
mkdir -p /opt/payment-api
cd /opt/payment-api

# Create a simple Node.js API (placeholder)
cat > index.js << 'EOF'
const http = require('http');
const port = 3000;

const server = http.createServer((req, res) => {{
  if (req.url === '/health') {{
    res.writeHead(200, {{'Content-Type': 'application/json'}});
    res.end(JSON.stringify({{ status: 'healthy', environment: '{environment_suffix}' }}));
  }} else {{
    res.writeHead(200, {{'Content-Type': 'application/json'}});
    res.end(JSON.stringify({{ message: 'Payment API', environment: '{environment_suffix}' }}));
  }}
}});

server.listen(port, '0.0.0.0', () => {{
  console.log('Payment API listening on port ' + port);
}});
EOF

# Create systemd service
cat > /etc/systemd/system/payment-api.service << 'SVCEOF'
[Unit]
Description=Payment API Service
After=network.target

[Service]
Type=simple
User=root
WorkingDirectory=/opt/payment-api
ExecStart=/usr/bin/node index.js
Restart=always
RestartSec=10

[Install]
WantedBy=multi-user.target
SVCEOF

# Start service
systemctl daemon-reload
systemctl enable payment-api
systemctl start payment-api
"""

        # Encode user_data in base64 for launch template
        user_data_encoded = base64.b64encode(user_data.encode('utf-8')).decode('utf-8')

        # Launch Template
        launch_template = LaunchTemplate(
            self,
            "launch_template",
            name=f"payment-api-lt-{environment_suffix}",
            image_id=ami.id,
            instance_type="t3.micro",
            vpc_security_group_ids=[api_security_group_id],
            user_data=user_data_encoded,
            iam_instance_profile=LaunchTemplateIamInstanceProfile(
                arn=instance_profile_arn,
            ),
            metadata_options=LaunchTemplateMetadataOptions(
                http_endpoint="enabled",
                http_tokens="required",
                http_put_response_hop_limit=1,
            ),
            monitoring=LaunchTemplateMonitoring(
                enabled=True,
            ),
            tags={
                "Name": f"payment-api-lt-{environment_suffix}",
            },
        )

        # Auto Scaling Group
        self._asg = AutoscalingGroup(
            self,
            "asg",
            name=f"payment-api-asg-{environment_suffix}",
            min_size=2,
            max_size=6,
            desired_capacity=2,
            vpc_zone_identifier=private_subnet_ids,
            target_group_arns=[self._target_group.arn],
            health_check_type="ELB",
            health_check_grace_period=300,
            launch_template=AutoscalingGroupLaunchTemplate(
                id=launch_template.id,
                version="$Latest",
            ),
            tag=[
                AutoscalingGroupTag(
                    key="Name",
                    value=f"payment-api-{environment_suffix}",
                    propagate_at_launch=True,
                ),
                AutoscalingGroupTag(
                    key="Environment",
                    value=environment_suffix,
                    propagate_at_launch=True,
                ),
            ],
        )

    @property
    def alb_dns_name(self) -> str:
        """Return ALB DNS name."""
        return self._alb.dns_name

    @property
    def alb_arn(self) -> str:
        """Return ALB ARN."""
        return self._alb.arn

    @property
    def alb_arn_suffix(self) -> str:
        """Return ALB ARN suffix."""
        return self._alb.arn_suffix

    @property
    def alb_target_group_arn_suffix(self) -> str:
        """Return target group ARN suffix."""
        return self._target_group.arn_suffix

    @property
    def asg_name(self) -> str:
        """Return Auto Scaling Group name."""
        return self._asg.name
